geometric/neb.py

```diff
@@ -0,0 +1,2289 @@
+    ForceRebuild = False
+    LastForce = 0
+    # How many steps since we have checked the coordinate system
+    CoordCounter = 0
+    # History of chains, including the current one
+    c_hist = [chain]
+    # == Enter the main optimization loop
+    optCycle = 0
+    LastRebuild = 1
+    respaced = False
+    Quality = None
+    while True:
+
+        # ======================================================#
+        # | At the top of the loop, our coordinates, energies, |#
+        # | gradients, and Hessians are synchronized.          |#
+        # ======================================================#
+
+        optCycle += 1
+        logger.info("\n-=# Chain optimization cycle %i #=- \n" % (optCycle))
+
+        # =======================================#
+        # |    Obtain an optimization step      |#
+        # |    and update Cartesian coordinates |#
+        # =======================================#
+
+        logger.info("Time since last ComputeChain: %.3f s \n" % (time.time() - t0))
+
+        (chain_prev, chain, expect, expectG, ForceRebuild, LastForce, Y_prev, GW_prev, GP_prev, respaced, optCycle) \
+        = takestep(c_hist, chain, optCycle, LastForce, ForceRebuild, trust, Y, GW, GP, HW, HP, None)
+
+        chain.ComputeChain(cyc=optCycle)
+        t0 = time.time()
+        # ----------------------------------------------------------
+        chain, Y, GW, GP, HW, HP, c_hist, Quality = compare(chain_prev, chain, ThreHQ, ThreLQ, GW, HW, HP, respaced,
+            optCycle, expect, expectG, trust, trustprint, params.avgg, params.maxg, Quality)
+
+        if respaced:
+            chain.SaveToDisk(fout="chain_%04i.xyz" % optCycle)
+            continue
+        chain.SaveToDisk(fout="chain_%04i.xyz" % optCycle)
+
+        # =======================================#
+        # |    Check convergence criteria       |#
+        # =======================================#
+
+        if converged(chain.maxg, chain.avgg, params.avgg, params.maxg, optCycle, params.maxcyc):
+            break
+
+        # =======================================#
+        # |  Adjust Trust Radius / Reject Step  |#
+        # =======================================#
+
+        chain, trust, trustprint, Y, GW, GP, good = qualitycheck(chain_prev, chain,trust, Quality, ThreLQ, ThreRJ,
+            ThreHQ, Y, GW, GP, Y_prev, GW_prev, GP_prev, params.tmax)
+        if not good:
+            continue
+        c_hist.append(chain)
+        c_hist = c_hist[-params.history :]
+
+        # =======================================#
+        # |      Update the Hessian Matrix      |#
+        # =======================================#
+
+        chain, Y, GW, GP, HP, HW, Y_prev, GP_prev, GW_prev, _ = updatehessian(chain_prev, chain, HP, HW, Y, Y_prev, GW,
+                                                                 GW_prev, GP, GP_prev, LastForce, params, None)
+    return chain, optCycle
+
+
+class nullengine(object):
```

Comment:
Perhaps we could put the QCFractal specific functions into a separate file, because this file is getting to be quite large.

Reply:
Done.